miniapps/meshing/tmop-fitting.cpp

```diff
@@ -0,0 +1,897 @@
+                  "terminate when max surface fitting error is below this limit");
+   args.AddOption(&adapt_marking, "-marking", "--adaptive-marking", "-no-amarking",
+                  "--no-adaptive-marking",
+                  "Enable or disable adaptive marking surface fitting.");
+   args.AddOption(&surf_bg_mesh, "-sbgmesh", "--surf-bg-mesh",
+                  "-no-sbgmesh","--no-surf-bg-mesh", "Use background mesh for surface fitting.");
+   args.AddOption(&comp_dist, "-dist", "--comp-dist",
+                  "-no-dist","--no-comp-dist", "Compute distance from 0 level set or not.");
+   args.AddOption(&surf_ls_type, "-slstype", "--surf-ls-type",
+                  "1 - Circle (DEFAULT), 2 - Squircle, 3 - Butterfly.");
+   args.AddOption(&marking_type, "-smtype", "--surf-marking-type",
+                  "1 - Interface (DEFAULT), 2 - Boundary attribute.");
+   args.AddOption(&mod_bndr_attr, "-mod-bndr-attr", "--modify-boundary-attribute",
+                  "-fix-bndr-attr", "--fix-boundary-attribute",
+                  "Change boundary attribue based on alignment with Cartesian axes.");
+   args.AddOption(&material, "-mat", "--mat",
+                  "-no-mat","--no-mat", "Use default material attributes.");
+   args.AddOption(&mesh_node_ordering, "-mno", "--mesh_node_ordering",
+                  "Ordering of mesh nodes."
+                  "0 (default): byNodes, 1: byVDIM");
+   args.AddOption(&amr_iters, "-amriter", "--amr-iter",
+                  "Number of amr iterations on background mesh");
+   args.Parse();
+   if (!args.Good())
+   {
+      if (myid == 0) { args.PrintUsage(cout); }
+      return 1;
+   }
+   if (myid == 0) { args.PrintOptions(cout); }
+
+   Device device(devopt);
+   if (myid == 0) { device.Print();}
+
+   // 3. Initialize and refine the starting mesh.
+   Mesh *mesh = new Mesh(mesh_file, 1, 1, false);
+   for (int lev = 0; lev < rs_levels; lev++)
+   {
+      mesh->UniformRefinement();
+   }
+   const int dim = mesh->Dimension();
+
+   // Define level-set coefficient
+   FunctionCoefficient *ls_coeff = NULL;
+   if (surf_ls_type == 1) //Circle
+   {
+      ls_coeff = new FunctionCoefficient(circle_level_set);
+   }
+   else if (surf_ls_type == 2) // reactor
+   {
+      ls_coeff = new FunctionCoefficient(reactor);
+   }
+   else if (surf_ls_type == 6) // 3D shape
+   {
+      ls_coeff = new FunctionCoefficient(csg_cubecylsph);
+   }
+   else
+   {
+      MFEM_ABORT("Surface fitting level set type not implemented yet.")
+   }
+
+   ParMesh *pmesh = new ParMesh(MPI_COMM_WORLD, *mesh);
+   delete mesh;
+   for (int lev = 0; lev < rp_levels; lev++) { pmesh->UniformRefinement(); }
+
+   // 4. Setup background mesh for surface fitting
+   ParMesh *pmesh_surf_fit_bg = NULL;
+   if (surf_bg_mesh)
+   {
+      Mesh *mesh_surf_fit_bg = NULL;
+      if (dim == 2)
+      {
+         mesh_surf_fit_bg =
+            new Mesh(Mesh::MakeCartesian2D(4, 4, Element::QUADRILATERAL, true));
```

Comment:
If this is run on many MPI, some tasks will be left empty initially, and they will stay empty after AMR. We can make the initial #elements depend on the #tasks.

Reply:
I rebalance after AMR so that should avoid the issue of empty tasks?